sandbox/libs/authn/src/main/java/org/opensearch/authn/Permission.java

```diff
@@ -0,0 +1,60 @@
+/*
+ * Copyright OpenSearch Contributors
+ * SPDX-License-Identifier: Apache-2.0
+ */
+
+package org.opensearch.authn;
+
+/**
+ * A permission of OpenSearch internal resources
+ *
+ * Example "opensearch.indexing.index.create"
+ *
+ * @opensearch.experimental
+ */
+public class Permission {
+
+    private final static String PERMISSION_DELIMITER = "\\.";
+
+    public final static String INVALID_CHARACTERS = ":"; // This is a placeholder for what may want to be banned
+
+    public final String[] permissionChunks;
+
+    public final String permissionString;
+
+    public Permission(final String permission) {
+        try {
+            this.permissionString = permission;
+            this.permissionChunks = permission.split(PERMISSION_DELIMITER);
+            if (!permissionIsValidFormat()) {
```

Comment:
This uses side effects to store information into the class and then read it later.  Seems unneeded when a validator function could accept a string and accept/reject it

Reply:
I agree with this analysis, take a read https://softwareengineering.stackexchange.com/questions/15269/why-are-side-effects-considered-evil-in-functional-programming

Reply:
I can move this outside of the constructor. I think that the valid format call would be best placed outside of the Permission class as a result however. The intention is to make it so that invalid permissions are never fully created. Which is why the formatting check is called inside the constructor since it will throw an exception on the incorrect formatting and then the Permission that is invalid will never be created in the first place. 

If we remove this from the constructor I think we should then move the formatting check into the handler class so that the handler could first check that a permission is correctly formatted and then assuming it is create it. It cannot do this if the function stays a method of the permission class since you would need to construct the permission before checking it.

Reply:
Note; if you don't want to throw an exception in a ctor you could always make the constructor private, but then expose a static `createInstance(...)` that effectively calls the constructor after verifying the object, or use a builder pattern [link](https://www.geeksforgeeks.org/builder-design-pattern/)

Reply:
Sounds good.